src/main/java/com/maxmind/db/Decoder.java

```diff
@@ -220,65 +207,257 @@ static int decodeInteger(ByteBuffer buffer, int base, int size) {
+            throw new DeserializationException();
+        }
+
+        Class<?> elementClass = cls.getComponentType();
+
+        Object array = Array.newInstance(elementClass, size);
 
-        List<JsonNode> array = new ArrayList<>(size);
         for (int i = 0; i < size; i++) {
-            JsonNode r = this.decode();
-            array.add(r);
+            Object e = this.decode(elementClass);
+            Array.set(array, i, e);
+        }
+
+        return array;
+    }
+
+    private <T> Object decodeMap(int size, Class<T> cls)
+            throws IOException,
+                   InstantiationException,
+                   IllegalAccessException,
+                   InvocationTargetException {
+        // We look for Object.class because if we're decoding a nested map into
+        // a Map, the second level is an Object.class. See Object.class in
+        // decodeMapIntoMap().
+        if (cls.equals(Map.class) || cls.equals(Object.class)) {
+            return this.decodeMapIntoMap(size);
         }
 
-        return new ArrayNode(OBJECT_MAPPER.getNodeFactory(), Collections.unmodifiableList(array));
+        return this.decodeMapIntoObject(size, cls);
     }
 
-    private JsonNode decodeMap(int size) throws IOException {
-        int capacity = (int) (size / 0.75F + 1.0F);
-        Map<String, JsonNode> map = new HashMap<>(capacity);
+    private HashMap<String, Object> decodeMapIntoMap(int size)
+            throws IOException,
+                   InstantiationException,
+                   IllegalAccessException,
+                   InvocationTargetException {
+        HashMap<String, Object> map = new HashMap<>();
 
         for (int i = 0; i < size; i++) {
-            String key = this.decode().asText();
-            JsonNode value = this.decode();
+            String key = (String) this.decode(String.class);
+
+            // Ideally we'd know the value's type to pass in, but I am having
+            // trouble retrieving it. I am not sure it is possible.
+            Object value = this.decode(Object.class);
+
             map.put(key, value);
         }
 
-        return new ObjectNode(OBJECT_MAPPER.getNodeFactory(), Collections.unmodifiableMap(map));
+        return map;
+    }
+
+    private <T> Object decodeMapIntoObject(int size, Class<T> cls)
+            throws IOException,
+                   InstantiationException,
+                   IllegalAccessException,
+                   InvocationTargetException {
+        Constructor constructor = this.findConstructor(cls);
+
+        Class<?>[] parameterTypes = constructor.getParameterTypes();
+
+        Map<String, Integer> parameterIndexes = new HashMap<>();
+        Annotation[][] annotations = constructor.getParameterAnnotations();
+        for (int i = 0; i < constructor.getParameterCount(); i++) {
+            String parameterName = this.getParameterName(cls, i, annotations[i]);
+            parameterIndexes.put(parameterName, i);
+        }
```

Comment:
This seems like something we could potentially cache. It shouldn't change at runtime.